js/MediaStream.ts

```diff
@@ -1,43 +1,56 @@
-/**
- * Expose the MediaStream class.
- */
-module.exports = MediaStream;
-
 /**
  * Spec: http://w3c.github.io/mediacapture-main/#mediastream
  */
 
-/**
- * Dependencies.
- */
-var debug = require('debug')('iosrtc:MediaStream'),
-	exec = require('cordova/exec'),
-	EventTarget = require('./EventTarget'),
-	MediaStreamTrack = require('./MediaStreamTrack'),
-	/**
-	 * Local variables.
-	 */
+import debugBase from 'debug';
+import { EventTargetShim } from './EventTarget';
+import {
+	MediaStreamTrackShim,
+	MediaStreamTrackAsJSON,
+	originalMediaStreamTrack
+} from './MediaStreamTrack';
 
-	// Dictionary of MediaStreams (provided via getMediaStreams() class method).
-	// - key: MediaStream blobId.
-	// - value: MediaStream.
-	mediaStreams;
+const exec = require('cordova/exec'),
+	debug = debugBase('iosrtc:MediaStream');
+
+// Dictionary of MediaStreams (provided via getMediaStreams() class method).
+// - key: MediaStream blobId.
+// - value: MediaStream.
+export const mediaStreams: { [blobId: string]: MediaStreamShim } = {};
 
 // TODO longer UUID like native call
 // - "4021904575-2849079001-3048689102-1644344044-4021904575-2849079001-3048689102-1644344044"
 function newMediaStreamId() {
 	return window.crypto.getRandomValues(new Uint32Array(4)).join('-');
 }
 
-// Save original MediaStream
-var originalMediaStream = window.MediaStream || window.Blob;
-//var originalMediaStream = window.Blob;
-var originalMediaStreamTrack = MediaStreamTrack.originalMediaStreamTrack;
+// Save original MediaStream or Blob as fallback in older iOS versions
+export const originalMediaStream =
+	(window.MediaStream as { new (): MediaStream } | undefined) || window.Blob;
 
-/**
- * Expose the MediaStream class.
- */
-function MediaStream(arg, id) {
+export interface MediaStreamAsJSON {
+	id: string;
+	audioTracks: { [id: string]: MediaStreamTrackAsJSON };
+	videoTracks: { [id: string]: MediaStreamTrackAsJSON };
+}
+
+interface AddTrackEvent {
+	type: 'addtrack';
+	track: MediaStreamTrackAsJSON;
+}
+
+interface RemoveTrackEvent {
+	type: 'removetrack';
+	track: MediaStreamTrackAsJSON;
+}
+
+type MediaStreamEvent = AddTrackEvent | RemoveTrackEvent;
+
+export const MediaStreamNativeShim = (function (
```

Comment:
This file ended up a little different than the rest.  I couldn't create a single `Shim` class because we need to return a `Blob` or native `MediaStream` from the constructor, which is not possible with a class.  Instead, the main export is a `function`, and it assigns all the properties from the `Shim` class onto the returned instance.  This should keep compatibility with what was here before.  I did try it in the sample app with the `MediaStream` stream available, and with it unavailable (falling back to `Blob`).

Reply:
Yep, that why Typescript never landed on this lib, it's unnecessary and just make the code more complicated to achieve something that JavaScript pure as no issue to achieve.

I will only land Typescript is fully working like previously regarding MediaStream SHIM.

Reply:
Makes sense.  I think I was able to strike the balance between making it work before, while getting the benefits of typescript.  All the other files turned out much nicer than pure JS by using straight classes.  Based on my testing, what I have here should work exactly like it did before.

Reply:
Makes sense.  I think I was able to strike the balance between making it work like it did before, while getting the benefits of typescript.  All the other files turned out much nicer than pure JS by using straight classes.  Based on my testing, what I have here should work exactly like it did before.

Reply:
Cool.

I will make final review this weekend and backport issue that you may have found in 6.0.17 before merging on master for 6.1.0